### PR TITLE
#163426494 implement-follow-unfollow-feature

### PR DESCRIPTION
**Pull request by biolabalo**
#### What does this PR do?
It creates API endpoints to
- Allow user follow an existing user
- Allow user unfollow a currently followed user
- Get all followers
- Get all users I follow
#### Description of Task to be completed?
Have the following endpoints working:

Follow/Unfollow user : `POST   /profiles/:username/follow`
Get all followers : `GET /profiles/users/following`
Get following : `GET /profiles/users/followers`

#### How should this be manually tested?
Clone this repo, cd into it and run npm install.
 **With test suite:**
     - Run `npm test`

 **Using postman:**
     - Run `npm run dev`,
     - With postman, send requests to `localhost:3000/<api url>` with the relevant HTTP methods and request payloads.

#### What are the relevant pivotal tracker stories?
[#163426494](https://www.pivotaltracker.com/story/show/163426494)

### CODE REVIEW

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by bobbylemm — DISMISSED** [review]
Apart from the reduction in code coverage, this looks good to me

File: server/controllers/followController.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import models from '../db/models';
+import errorMessage from '../helpers/errorHelpers';
+import followhelper from '../helpers/followAndUnFollow';
+
+const { Users, Follows } = models;
+
+/**
+ * @description Contains all follow and unfollow functionalities
+ * @export
+ * @class followAndUnfollow
```
Comment (primuse) [inline server/controllers/followController.js]: This is a wrong class name

File: server/controllers/followController.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ */
+class FollowController {
+  /**
+ *
+ * @description Method to follow user
+ * @static
+ * @param {object} req client request
+ * @param {object} res erver response
```
Comment (primuse) [inline server/controllers/followController.js]: `server` response. You're missing an `s`

File: server/controllers/followController.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ * @returns {Object} server response object
+ * @param {object} next server response
+ * @memberof FollowController
+ */
+  static async followUser(req, res, next) {
+    try {
+      const heplerResult = await followhelper(req, res, 'followUser');
```
Comment (primuse) [inline server/controllers/followController.js]: I believe this should be `helperResult` not `heplerResult`

File: server/controllers/followController.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+
+
+  /**
+ *
+ * @description Method to fetch all users who I follow
+ * @static
+ * @param {object} req client request
+ * @param {object} res erver response
```
Comment (primuse) [inline server/controllers/followController.js]: `server` response. You're missing an `s`

File: server/controllers/followController.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      next(error);
+    }
+  }
+
+  /**
+ *
+ * @description Method to fetch all users who follow me
+ * @static
+ * @param {object} req client request
+ * @param {object} res erver response
```
Comment (primuse) [inline server/controllers/followController.js]: Same as above.

File: server/controllers/followController.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+ *
+ * @description Method to fetch all users who I follow
+ * @static
+ * @param {object} req client request
+ * @param {object} res erver response
+ * @returns {Object} server response object
+ * @param {object} next server response
+ * @memberof followAndUnfollow
```
Comment (primuse) [inline server/controllers/followController.js]: This should be `FollowController` as that's the class name

File: server/controllers/followController.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+  /**
+ *
+ * @description Method to fetch all users who follow me
+ * @static
+ * @param {object} req client request
+ * @param {object} res erver response
+ * @returns {Object} server response object
+ * @param {object} next server response
+ * @memberof followAndUnfollow
```
Comment (primuse) [inline server/controllers/followController.js]: This should be `FollowController` as that's the class name

File: server/tests/app/followAndUnfollow.test.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+  it('should notify users without followers that they do not have followers', () => request(app)
+    .get('/api/v1/profiles/users/followers')
+    .set('x-auth-token', tokenForBalogun)
+    .then((res) => {
+      expect(res.status).to.equal(200);
+      expect(res.body.message).to.equal('You currently do not have any followers');
+    }));
+
+  it('should return a list of all user I follow', () => request(app)
```
Comment (primuse) [inline server/tests/app/followAndUnfollow.test.js]: User should be plural... `users`

**Review by ozone4real — DISMISSED** [review]
LGTM

File: server/controllers/followController.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+ */
+class FollowController {
+  /**
+ *
+ * @description Method to follow user
+ * @static
+ * @param {object} req client request
+ * @param {object} res server response
+ * @returns {Object} server response object
+ * @param {object} next server response
```
Comment (omenkish) [inline server/controllers/followController.js]: next is a function not an object

**Review by omenkish — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by EfeAgare — DISMISSED** [review]

**Review by bobbylemm — DISMISSED** [review]
LGTM

**Review by omenkish — DISMISSED** [review]
Looks good.

**Review by omenkish — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

File: server/controllers/followController.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+ * @static
+ * @param {object} req client request
+ * @param {object} res server response
+ * @returns {Object} server response object
+ * @param  {Function} next passes control to the next middleware
+ * @memberof FollowController
+ */
+  static async follow(req, res, next) {
+    try {
+      const helperResult = await followhelper(req, res, 'followUser');
```
Comment (p-funky) [inline server/controllers/followController.js]: The `followUser` param can be done away with.

**Review by p-funky — APPROVED** [review]

**Review by bobbylemm — APPROVED** [review]
LGTM

**Pull request merged**